core/trino-main/src/main/java/io/trino/operator/PartitionedOutputOperator.java

```diff
@@ -398,9 +394,18 @@ public long getRetainedSizeInBytes()
             return sizeInBytes;
         }
 
-        public PartitionedOutputInfo getInfo()
+        public Supplier<PartitionedOutputInfo> getOperatorInfoSupplier()
+        {
+            return createPartitionedOutputOperatorInfoSupplier(rowsAdded, pagesAdded, outputBuffer);
+        }
+
+        private static Supplier<PartitionedOutputInfo> createPartitionedOutputOperatorInfoSupplier(AtomicLong rowsAdded, AtomicLong pagesAdded, OutputBuffer outputBuffer)
         {
-            return new PartitionedOutputInfo(rowsAdded.get(), pagesAdded.get(), outputBuffer.getPeakMemoryUsage());
+            // Must be a separate static method to avoid embedding references to "this" in the supplier
```

Comment:
interesting. Won't java optimize it even supplier does not use this?

Reply:
The code we hope get out of the generated lambda is something equivalent to this:
```
private static class InfoSupplier implements Supplier<PartitionedOutputInfo> {
  private final AtomicLong rowsAdded;
  private final AtomicLong pagesAdded;
  private final OutputBuffer outputBuffer;
  @Override
  public PartitionedOutputInfo get() {
    return new PartitionedOutputInfo(rowsAdded.get(), pagesAdded.get(), outputBuffer.getPeakMemoryUsage());
  }
}
```

But if the lambda is created in an instance scoped method, it would be equally valid for the lambda to be expanded into something that would retain the whole operator instance like this:
```
private static class InfoSupplier implements Supplier<PartitionedOutputInfo> {
  private final PartitionedOutputOperator outer$this;
  @Override
  public PartitionedOutputInfo get() {
    return new PartitionedOutputInfo(outer$this.rowsAdded.get(), outer$this.pagesAdded.get(), outer$this.outputBuffer.getPeakMemoryUsage());
  }
}
```

The JVM language spec doesn't seem to mandate which behavior to use, except that the second (undesirable) option is permissable for lambdas defined within an instance scoped method. I wouldn't be surprised if there was a heuristic for which formulation to use based on the number of captured field references. In any case, we can ensure that the undesirable form won't be used by defining the lambda inside of a static scope.